File: 프로그래머스/lv3/지준혁/길 찾기 게임.cpp
#include <string>
#include <vector>
#include <algorithm>
#include <iostream>
#include <tuple>

using namespace std;

int lc[10'004];
int rc[10'004];

void setParent(int pa_idx, int ch_idx, int ch_x, vector<vector<int>>& nodeinfo) {
    
    int pa_x = nodeinfo[pa_idx - 1][0];
        
    if (ch_x < pa_x) {
        if (lc[pa_idx] == 0) {
            lc[pa_idx] = ch_idx;
        } 
        else {
            setParent(lc[pa_idx], ch_idx, ch_x, nodeinfo);
        }
    } else {
        if (rc[pa_idx] == 0) {
            rc[pa_idx] = ch_idx;
        } 
        else {
            setParent(rc[pa_idx], ch_idx, ch_x, nodeinfo);
        }
    }
}

void preorder(int idx, vector<int>& ret) {
    if (idx == 0) return;
    
    ret.push_back(idx);
    preorder(lc[idx], ret);
    preorder(rc[idx], ret);
}

void postorder(int idx, vector<int>& ret) {
    if (idx == 0) return;
    
    postorder(lc[idx], ret);
    postorder(rc[idx], ret);
    ret.push_back(idx);
}

vector<vector<int>> solution(vector<vector<int>> nodeinfo) {
    vector<tuple<int, int, int>> nodes;

    for (int i = 0; i < nodeinfo.size(); ++i) {
        nodes.push_back({i + 1, nodeinfo[i][0], nodeinfo[i][1]});
    }

    sort(nodes.begin(), nodes.end(), [](auto& a, auto& b) {
        if (get<2>(a) != get<2>(b)) {
            return get<2>(a) > get<2>(b);
        }
        return get<1>(a) < get<1>(b);
    });

    
    int root_idx = get<0>(nodes[0]);
    for (int i = 1; i < nodes.size(); ++i) {
        setParent(root_idx, get<0>(nodes[i]), get<1>(nodes[i]), nodeinfo);
    }

    vector<vector<int>> ans;
    vector<int> pre, post;
    
    preorder(root_idx, pre);
    postorder(root_idx, post);

    ans.push_back(pre);
    ans.push_back(post);

    return ans;
}

File: 프로그래머스/lv3/최원준/길 찾기 게임.py
'''
1. 아이디어 :

2. 시간복잡도 :
    O(  )
3. 자료구조 :

'''

import sys
sys.setrecursionlimit(10000)
def solution(nodeinfo):
    nodes = [(x, y, i+1) for i, (x, y) in enumerate(nodeinfo)]
    nodes.sort(key=lambda x: (-x[1], x[0]))
    print(nodes)
    pre_order = []
    post_order = []

    def dfs(nodes):
        if not nodes:
            return None

        root = nodes[0]
        left_subtree = [node for node in nodes if node[0] < root[0]]
        right_subtree = [node for node in nodes if node[0] > root[0]]

        pre_order.append(root[2])  # 전위 순회: 루트 방문
        dfs(left_subtree)          # 왼쪽 서브트리 방문
        dfs(right_subtree)         # 오른쪽 서브트리 방문
        post_order.append(root[2]) # 후위 순회: 루트 방문

        return root

    dfs(nodes)

    return [pre_order, post_order]